llvm/include/llvm/CodeGen/GlobalISel/LegalizerInfo.h

```diff
@@ -220,7 +220,8 @@ struct TypePairAndMemDesc {
            Align >= Other.Align &&
            // FIXME: This perhaps should be stricter, but the current legality
            // rules are written only considering the size.
-           MemTy.getSizeInBits() == Other.MemTy.getSizeInBits();
```

Comment:
Did this really cause a problem? It looks like `TypeSize::operator==` should not cause a crash for scalable vectors:

```
constexpr bool operator==(const FixedOrScalableQuantity &RHS) const {
  return Quantity == RHS.Quantity && Scalable == RHS.Scalable;
}
```

Reply:
Comparing the min value is losing the scalable bit. Why a fixed size and scalable size should not be compatible right?

Reply:
> Comparing the min value is losing the scalable bit. Why a fixed size and scalable size should not be compatible right?

Should scalable `nxv2s8` be compatible with `s8`? What about `nxv2s8` and `s8`? I'm a bit unsure about the logic behind

Reply:
If you have an s8 its type size in bits will be `(ElementCount::Fixed(1), 8 bits)` If you have an nxv2s8 its type size in bits will be `(ElementCount::Sclable(2), 8)`. `(ElementCount::Fixed(1), 8) != (ElementCount::Sclable(2), 8)`

Reply:
Right, `(ElementCount::Scalable(2), 8) != (ElementCount::Fixed(1), 8)`, but does that mean they are incompatible? 
Since `(ElementCount::Scalable(2), 8)` can hold at least two `s8` types, it'll definitely be a multiple of `(ElementCount::Fixed(1), 8)`. So it seems to me `(ElementCount::Scalable(2), 8)` can hold/be "compatible" with `(ElementCount::Fixed(1), 8)`, if it makes sense.

Reply:
I think that it would be a good idea to restrict the types to match here. If we want to do loads across different types then we should have some opcode that "adjusts the type" like a bitcast:

```
%0(vscale x 2 x s8) = G_IMPLICIT_DEF
%1(vscale x 1 x s16 = G_BITCAST %0
...
```

Reply:
It's true we can do BIT_CAST if `ScalableEltTy != FixedEltTy` && `MinNumElts * EltTy == FixedEltTy` (2 * 8 == 16 in your example).
I was wondering what if `ScalableEltTy == FixedEltTy` and `MinNumElts > 1`? In these cases, the scalable vector is the "superset" of the fixed one so they can be compatible in the sense of fitting one into another. Am I understanding it correctly?

Reply:
We don't currently support a G_BITCAST between a scalar and a vector at the moment. I was using it as an example to show that if the types do not obviously match, then maybe it is a better idea to explicitly convert them into a form that does match using an instruction.

In the case of `ScalableEltTy == ScalarEltTy and MinNumElts > 1` I am suggesting that we disallow the load because the TypeSize does not match, and instead rely on an instruction to get us in a form where type size does match. My point is that it is possible but we should be explicit in making it possible. Vectors and scalars use different registers on RISC-V and it is likely that that "convert instruction" will make it easier to select a transfer from vector to scalar or scalar to vector if we need to do that.

Reply:
Make sense, thanks!
To answer your question:
> Did this really cause a problem? It looks like TypeSize::operator== should not cause a crash for scalable vectors: ...

The answer is yes. And the reason why I was changing to `MemTy.getSizeInBits().getKnownMinValue() == Other.MemTy.getSizeInBits().getKnownMinValue()` is because I have:
https://github.com/llvm/llvm-project/blob/87504bb181af547827d01574912962e14043e1ae/llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp#L213-L218
In this case, `MemTy` is `nxv1s8` and `Other.MemTy` is `s8`. Therefore, based on our discussion, they should be incompatible. And seems like the only solution to make them compatible is:
```cpp
.legalForTypesWithMemDesc({{s32, p0, s8, 8},
                                                     {s32, p0, s16, 16},
                                                     {s32, p0, s32, 32},
                                                     {p0, p0, sXLen, XLen},
                                                     {nxv1s8, p0, nxv1s8, 8}});
```
But this doesn't look right to me..

Reply:
The third parameter definitely needs to be a scalable type to match true size of nxv1s8.

AArch64 has this for vectors, but they don't support scalable vectors yet. The third parameter is a scalar the same total size as the vector
```
      .legalForTypesWithMemDesc({{s8, p0, s8, 8},                                
                                 {s16, p0, s16, 8},                              
                                 {s32, p0, s32, 8},                              
                                 {s64, p0, s64, 8},                              
                                 {p0, p0, s64, 8},                               
                                 {s128, p0, s128, 8},                            
                                 {v8s8, p0, s64, 8},                             
                                 {v16s8, p0, s128, 8},                           
                                 {v4s16, p0, s64, 8},                            
                                 {v8s16, p0, s128, 8},                           
                                 {v2s32, p0, s64, 8},                            
                                 {v4s32, p0, s128, 8},                           
                                 {v2s64, p0, s128, 8}})   
```

So `{nxv1s8, p0, s8, 8}` seems wrong to me, but I'm not sure what we can put there.

Reply:
I guess since `isCompatible` just checks the size, I guess `{v16s8, p0, s128, 8}` or `{v16s8, p0, v16s8, 8}` are equivalent for AArch64?

So using `{nxv1s8, p0, nxv1s8, 8}` would be correct for RISC-V.